<a href="https://colab.research.google.com/github/m-dwaipayan/CDB22DW022-Dwaipayan/blob/main/Python_NumPy_practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**NumPy**

In [1]:
# Importing numPy
import numpy as np

In [2]:
# Defining a python list containing age of 6 people
ages_list = [10,5,8,32,65,43]
print(ages_list)

[10, 5, 8, 32, 65, 43]


In [3]:
# Instantiating a NumPy array object using np.array()
ages = np.array(ages_list)
print(type(ages))
print(ages)

<class 'numpy.ndarray'>
[10  5  8 32 65 43]


In [5]:
# Printing the size and shape of the array ages
print(ages)
print("Size:\t",ages.size)
print("Shape:\t",ages.shape)

[10  5  8 32 65 43]
Size:	 6
Shape:	 (6,)


In [6]:
# Creating an array of zeros
zeroArr = np.zeros(5)
print(zeroArr)

[0. 0. 0. 0. 0.]


Multi-Dimensional

In [7]:
# Defining a list of the weights of these 6 individuals
weight_list = [32,18,26,60,55,65]

In [8]:
# Creating an array with both age and weight of the 6 individuals
people = np.array([ages_list,weight_list])
print("People:\t",people)

# Printing the size and shape
print("Size:\t",people.size)
print("Shape:\t",people.shape)

People:	 [[10  5  8 32 65 43]
 [32 18 26 60 55 65]]
Size:	 12
Shape:	 (2, 6)


In [9]:
# Changing the shape and printing
people = people.reshape(12,1)
print("People:\t",people)
print("Size:\t",people.size)
print("Shape:\t",people.shape)

People:	 [[10]
 [ 5]
 [ 8]
 [32]
 [65]
 [43]
 [32]
 [18]
 [26]
 [60]
 [55]
 [65]]
Size:	 12
Shape:	 (12, 1)


In [10]:
# Changing the shape and printing
people = people.reshape(4,3)
print("People:\t",people)
print("Size:\t",people.size)
print("Shape:\t",people.shape)

People:	 [[10  5  8]
 [32 65 43]
 [32 18 26]
 [60 55 65]]
Size:	 12
Shape:	 (4, 3)


In [11]:
# Changing the shape and printing
people = people.reshape(3,4)
print("People:\t",people)
print("Size:\t",people.size)
print("Shape:\t",people.shape)

People:	 [[10  5  8 32]
 [65 43 32 18]
 [26 60 55 65]]
Size:	 12
Shape:	 (3, 4)


In [12]:
# Changing the shape and printing
people = people.reshape(6,2)
print("People:\t",people)
print("Size:\t",people.size)
print("Shape:\t",people.shape)

People:	 [[10  5]
 [ 8 32]
 [65 43]
 [32 18]
 [26 60]
 [55 65]]
Size:	 12
Shape:	 (6, 2)


In [13]:
# Changing the shape and printing
people = people.reshape(1,12)
print("People:\t",people)
print("Size:\t",people.size)
print("Shape:\t",people.shape)

People:	 [[10  5  8 32 65 43 32 18 26 60 55 65]]
Size:	 12
Shape:	 (1, 12)


Exercise

In [14]:
# Generating 1D numpy array with values [7,9,65,33,85,99]
new_list = [7,9,65,33,85,99]
newArr = np.array(new_list)
print("Array:\t",newArr)
print("Size:\t",newArr.size)
print("Shape:\t",newArr.shape)

Array:	 [ 7  9 65 33 85 99]
Size:	 6
Shape:	 (6,)


In [15]:
# Generating a 2D NumPy array(matrix)
new_list1 = [1,2,4]
new_list2 = [2,3,0]
new_list3 = [0,5,1]
arr2D = np.array([new_list1,new_list2,new_list3]).reshape(3,3)
print("A=\t",arr2D)

A=	 [[1 2 4]
 [2 3 0]
 [0 5 1]]


In [16]:
# Reshaping the new array
A = arr2D.reshape(9,1)
print(A)

[[1]
 [2]
 [4]
 [2]
 [3]
 [0]
 [0]
 [5]
 [1]]


ARRAY GENERATION

In [20]:
# Creating an array using np.arange()
five_times_table = np.arange(0,55,5)
print(five_times_table)

[ 0  5 10 15 20 25 30 35 40 45 50]


In [21]:
# Creating an array using np.linspace()
five_spaced = np.linspace(0,50,11)
print(five_spaced)

[ 0.  5. 10. 15. 20. 25. 30. 35. 40. 45. 50.]


In [22]:
# Creating an array using np.repeat()
twoArr = np.repeat(2,10)
print(twoArr)

[2 2 2 2 2 2 2 2 2 2]


In [23]:
# Creating an identity matrix
identity_matrix = np.eye(6)
print(identity_matrix)

[[1. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 0. 1.]]


Operations


In [24]:
five_times_table

array([ 0,  5, 10, 15, 20, 25, 30, 35, 40, 45, 50])

In [26]:
# Returning an array with twice the elements
print("1:", 2 * five_times_table)

1: [  0  10  20  30  40  50  60  70  80  90 100]


In [27]:
# Returning an array with 10 added to each element
print("2:", 10 + five_times_table)

2: [10 15 20 25 30 35 40 45 50 55 60]


In [28]:
# Returning an array with 1 subtracted from each element
print("3:", five_times_table - 1)

3: [-1  4  9 14 19 24 29 34 39 44 49]


In [29]:
# Returning an array with each element divided by 5
print("4:", five_times_table/5)

4: [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10.]


In [30]:
# Returning an array with each element raised to the power of 2
print("5:", five_times_table ** 2)

5: [   0   25  100  225  400  625  900 1225 1600 2025 2500]


In [31]:
# Returning an array of boolean values if the elements are less than 20
print("6:", five_times_table < 20)

6: [ True  True  True  True False False False False False False False]


Speed Test

In [32]:
fives_list = list(range(0,5001,5))
fives_list

[0,
 5,
 10,
 15,
 20,
 25,
 30,
 35,
 40,
 45,
 50,
 55,
 60,
 65,
 70,
 75,
 80,
 85,
 90,
 95,
 100,
 105,
 110,
 115,
 120,
 125,
 130,
 135,
 140,
 145,
 150,
 155,
 160,
 165,
 170,
 175,
 180,
 185,
 190,
 195,
 200,
 205,
 210,
 215,
 220,
 225,
 230,
 235,
 240,
 245,
 250,
 255,
 260,
 265,
 270,
 275,
 280,
 285,
 290,
 295,
 300,
 305,
 310,
 315,
 320,
 325,
 330,
 335,
 340,
 345,
 350,
 355,
 360,
 365,
 370,
 375,
 380,
 385,
 390,
 395,
 400,
 405,
 410,
 415,
 420,
 425,
 430,
 435,
 440,
 445,
 450,
 455,
 460,
 465,
 470,
 475,
 480,
 485,
 490,
 495,
 500,
 505,
 510,
 515,
 520,
 525,
 530,
 535,
 540,
 545,
 550,
 555,
 560,
 565,
 570,
 575,
 580,
 585,
 590,
 595,
 600,
 605,
 610,
 615,
 620,
 625,
 630,
 635,
 640,
 645,
 650,
 655,
 660,
 665,
 670,
 675,
 680,
 685,
 690,
 695,
 700,
 705,
 710,
 715,
 720,
 725,
 730,
 735,
 740,
 745,
 750,
 755,
 760,
 765,
 770,
 775,
 780,
 785,
 790,
 795,
 800,
 805,
 810,
 815,
 820,
 825,
 830,
 835,
 840,
 845,
 8

In [33]:
five_times_table_lge = np.arange(0,5001,5)
five_times_table_lge

array([   0,    5,   10, ..., 4990, 4995, 5000])

In [34]:
%timeit five_times_table_lge + 5

1.25 µs ± 12.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [35]:
%timeit [e + 5 for e in fives_list]

59.4 µs ± 565 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


String Operations

In [36]:
words = np.array( ["ten","nine","eight","seven","six"])
print(np.isin(words,'e'))

print("e" in words)
["e" in word for word in words]

[False False False False False]
False


[True, True, True, True, False]

Transpose

In [37]:
# Transpose of a matrix
people.shape = (2,6)
print(people, "\n")
print(people.T)

[[10  5  8 32 65 43]
 [32 18 26 60 55 65]] 

[[10 32]
 [ 5 18]
 [ 8 26]
 [32 60]
 [65 55]
 [43 65]]


Data Types

In [38]:
# Obtaining the data type of an array
people.dtype

dtype('int64')

In [40]:
ages_with_strings = np.array([10,5,8,'32','65','43'])
ages_with_strings

# To obtain the data type
ages_with_strings.dtype

dtype('<U21')

In [42]:
ages_with_strings = np.array([10,5,8,'32','65','43'], dtype = 'int32')
ages_with_strings

# Data type
ages_with_strings.dtype

dtype('int32')

In [44]:
ages_with_strings = np.array([10,5,8,'32','65','43'])
print(ages_with_strings)

['10' '5' '8' '32' '65' '43']


In [45]:
ages_with_strings.dtype = 'int32'
print(ages_with_strings)

[49 48  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0 53  0  0
  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0 56  0  0  0  0  0
  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0 51 50  0  0  0  0  0  0  0
  0  0  0  0  0  0  0  0  0  0  0  0 54 53  0  0  0  0  0  0  0  0  0  0
  0  0  0  0  0  0  0  0  0 52 51  0  0  0  0  0  0  0  0  0  0  0  0  0
  0  0  0  0  0  0]


In [46]:
ages_with_strings.size

126

In [47]:
ages_with_strings.size/21

6.0

In [48]:
np.array([10,5,8,'32','65','43']).size


6

What happened here was when we tried to change the datatype to int32, the array size got multiplied by 21 and 120 zeros aappeared.

In [49]:
# Changing the data type using the astype()
ages_with_strings = np.array([10,5,8,'32','65','43'])
print(ages_with_strings)
print(ages_with_strings.astype('int32'))

['10' '5' '8' '32' '65' '43']
[10  5  8 32 65 43]


EXERCISE

In [50]:
arrString = np.array(['1','2','5','7','9'])
print(arrString)

['1' '2' '5' '7' '9']


In [53]:
arrString.dtype = 'float32'
print(arrString)

[6.9e-44 7.0e-44 7.4e-44 7.7e-44 8.0e-44]


In [54]:
arrString2 = arrString.T
print(arrString2)
print("Size:\t",arrString2.size)
print("Shape:\t",arrString2.shape)


[6.9e-44 7.0e-44 7.4e-44 7.7e-44 8.0e-44]
Size:	 5
Shape:	 (5,)


In [56]:
print(arrString.astype(bool))

[ True  True  True  True  True]


ARRAY SLICING OPERATION

In [57]:
five_times_table

array([ 0,  5, 10, 15, 20, 25, 30, 35, 40, 45, 50])

In [58]:
five_times_table[0]

0

In [59]:
five_times_table[-1]

50

In [60]:
five_times_table[:4]

array([ 0,  5, 10, 15])

In [61]:
five_times_table[4:]

array([20, 25, 30, 35, 40, 45, 50])

In [62]:
# for 2D arrays
print(people)
people[:3,:3]

[[10  5  8 32 65 43]
 [32 18 26 60 55 65]]


array([[10,  5,  8],
       [32, 18, 26]])

EXERCISE

In [64]:
# an array with 50 zeros
arrZero = np.zeros(50)
print(arrZero)

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0.]


In [65]:
# an array twenty 2s
arrTwos = np.repeat(2,20)
print(arrTwos)

[2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2]


In [67]:
# an array from 0 to 2 pi 
import math
arrPi = np.arange(0,2*math.pi,0.1)
print(arrPi)

[0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7
 1.8 1.9 2.  2.1 2.2 2.3 2.4 2.5 2.6 2.7 2.8 2.9 3.  3.1 3.2 3.3 3.4 3.5
 3.6 3.7 3.8 3.9 4.  4.1 4.2 4.3 4.4 4.5 4.6 4.7 4.8 4.9 5.  5.1 5.2 5.3
 5.4 5.5 5.6 5.7 5.8 5.9 6.  6.1 6.2]


In [68]:
# first five values 
arrPi[:5]

array([0. , 0.1, 0.2, 0.3, 0.4])

In [73]:
# last 3 values
arrPi[-3:]

array([6. , 6.1, 6.2])

In [75]:
# 4th to 7th value
arrPi[3:7]

array([0.3, 0.4, 0.5, 0.6])

In [76]:
# Reversing an array using flip()
reverse_five_times_table = np.flip(five_times_table)
reverse_five_times_table

array([50, 45, 40, 35, 30, 25, 20, 15, 10,  5,  0])

In [77]:
reverse_five_times_table = five_times_table[-1::-1]
print(reverse_five_times_table)
five_times_table

[50 45 40 35 30 25 20 15 10  5  0]


array([ 0,  5, 10, 15, 20, 25, 30, 35, 40, 45, 50])

In [78]:
# Every third element starting from 0
five_times_table[0::3]

array([ 0, 15, 30, 45])

EXERCISE

In [79]:
reverse_arrPi = np.flip(arrPi)
print(reverse_arrPi)

[6.2 6.1 6.  5.9 5.8 5.7 5.6 5.5 5.4 5.3 5.2 5.1 5.  4.9 4.8 4.7 4.6 4.5
 4.4 4.3 4.2 4.1 4.  3.9 3.8 3.7 3.6 3.5 3.4 3.3 3.2 3.1 3.  2.9 2.8 2.7
 2.6 2.5 2.4 2.3 2.2 2.1 2.  1.9 1.8 1.7 1.6 1.5 1.4 1.3 1.2 1.1 1.  0.9
 0.8 0.7 0.6 0.5 0.4 0.3 0.2 0.1 0. ]


In [81]:
reverse_arrPi[::4]

array([6.2, 5.8, 5.4, 5. , 4.6, 4.2, 3.8, 3.4, 3. , 2.6, 2.2, 1.8, 1.4,
       1. , 0.6, 0.2])

In [83]:
reverse_arrPi[-1::-2]

array([0. , 0.2, 0.4, 0.6, 0.8, 1. , 1.2, 1.4, 1.6, 1.8, 2. , 2.2, 2.4,
       2.6, 2.8, 3. , 3.2, 3.4, 3.6, 3.8, 4. , 4.2, 4.4, 4.6, 4.8, 5. ,
       5.2, 5.4, 5.6, 5.8, 6. , 6.2])

STATS

In [84]:
np.array([1.65432,5.98765]).round(2)

array([1.65, 5.99])

In [86]:
nums = np.arange(0,4,0.2555)

In [87]:
# Printing the minimum,maximum,sum,mean,median,variance,standard deviation 
print("min= ",np.min(nums).round(2))
print("max= ",np.max(nums).round(2))
print("sum= ",np.sum(nums).round(2))
print("mean= ",np.mean(nums).round(2))
print("median= ",np.median(nums).round(2))
print("var= ",np.var(nums).round(2))
print("std= ",np.std(nums).round(2))

min=  0.0
max=  3.83
sum=  30.66
mean=  1.92
median=  1.92
var=  1.39
std=  1.18


RANDOM

In [88]:
# Fair coin toss
flip = np.random.choice([0,1],10)
flip

array([0, 0, 0, 1, 0, 1, 1, 0, 0, 1])

In [90]:
np.random.rand(10,20,9)

array([[[0.67207647, 0.57471233, 0.61556312, ..., 0.9394185 ,
         0.58757329, 0.24989032],
        [0.0898844 , 0.28651162, 0.85214766, ..., 0.38849161,
         0.75271509, 0.63534583],
        [0.52406843, 0.42062967, 0.53363482, ..., 0.87796759,
         0.59779465, 0.64816714],
        ...,
        [0.68256601, 0.58495132, 0.71525985, ..., 0.66945964,
         0.27537693, 0.26286749],
        [0.72370156, 0.97596787, 0.07003584, ..., 0.43474393,
         0.22939469, 0.54969239],
        [0.75195815, 0.20397406, 0.00964109, ..., 0.82050619,
         0.90678741, 0.68114272]],

       [[0.74377519, 0.40502129, 0.08560368, ..., 0.18421008,
         0.52922382, 0.68239912],
        [0.1120986 , 0.53896036, 0.32170788, ..., 0.95912814,
         0.75061021, 0.44397288],
        [0.34030131, 0.02740837, 0.75592044, ..., 0.63515245,
         0.10695235, 0.94062556],
        ...,
        [0.26064785, 0.16177165, 0.39121772, ..., 0.33587518,
         0.78735106, 0.69864082],
        [0.5

In [92]:
# Datapoints from a normal distribution
mu,sigma = 0, 0.1 # mean and standard deviation
s = np.random.normal(mu,sigma,1000)
print(s)

[ 2.35861292e-02  1.74565129e-02  2.29264555e-02 -1.12223129e-01
 -5.70835968e-02 -1.37671228e-01  4.63643247e-02 -1.66250881e-01
 -5.04442789e-02 -1.25919650e-02 -7.20022827e-02 -7.03937307e-02
  3.05356725e-02  1.26952022e-01  6.36503263e-02  5.77136898e-02
  2.67408088e-02 -5.02575620e-02  7.80457893e-02  1.29290842e-02
 -2.15032733e-01 -1.54792725e-01  2.59684108e-01  1.10910151e-01
 -7.50763310e-02  1.36590121e-01  1.12529817e-01 -9.32781129e-02
  6.37552809e-02 -2.89826674e-02 -1.80862899e-01  3.86103256e-02
 -1.41508157e-01  1.45002405e-01  1.05654076e-01 -1.28561729e-02
 -6.81625629e-02  1.23029121e-01  7.93929071e-03 -5.70593261e-02
 -1.33179746e-01  6.40600920e-02 -3.21657160e-02 -1.44987653e-01
 -8.49019559e-02  6.24828790e-02 -3.00079606e-02 -1.23538895e-01
 -6.03721604e-02  1.73664646e-01  1.04836779e-01 -5.83743334e-03
  4.40238530e-02 -1.15816815e-01  8.50294705e-02 -4.46848784e-02
 -2.12891992e-01  1.31174983e-01  3.73014049e-02 -4.35590588e-02
 -1.08798617e-01  2.76965

EXERCISE

In [94]:
# Simulating a six-sided dice
die = np.random.choice([1,2,3,4,5,6],10)
print(die)

[3 5 3 4 2 6 5 4 2 2]


In [98]:
# Simulating a two_sided coin toss which is biased
coin = np.random.choice(['Head','Tail'],10,p=[0.67,0.33])
print(coin)

['Head' 'Head' 'Head' 'Head' 'Head' 'Tail' 'Head' 'Tail' 'Tail' 'Head']
